Add Container Insights getting start page, Prometheus and ECS Prometheus (#128)
aws-otel/aws-otel.github.io: src/docs/getting-started/container-insights/ecs-prometheus.mdx

---
title: 'Container Insights ECS Prometheus'
description:
    Container Insights ECS Prometheus discovers targets running on ECS EC2 and fargate.
    Workloads like NGINX, JMX comes with pre-built CloudWatch dashboard.
path: '/docs/getting-started/container-insights/ecs-prometheus'
---

import {Link} from "gatsby"

<p>
    GettingStarted / <Link to="/docs/getting-started/container-insights">Container Insights</Link> / <Link to="/docs/getting-started/container-insights/prometheus">Container Insights Prometheus</Link>
</p>

## Overview

In order to collect metrics from task running on ECS and send it to CloudWatch, you can either use a sidecar or run collector as ECS service for entire cluster.

> For using pre-built dashboard in ClouDWatch Container Insights Prometheus, you have to run a single collector for entire cluster to attach task and service metadata.

Reviewer: ```suggestion
To use pre-built dashboard in Amazon CloudWatch Container Insights Prometheus, you have to run a single collector for entire cluster to attach task and service metadata.
```

---
title: 'Container Insights ECS Prometheus'
description:
    Container Insights ECS Prometheus discovers targets running on ECS EC2 and fargate.
    Workloads like NGINX, JMX comes with pre-built CloudWatch dashboard.
path: '/docs/getting-started/container-insights/ecs-prometheus'
---

import {Link} from "gatsby"
import ecssdviewlog from "assets/img/docs/gettingStarted/container-insights/ecssd-view-log.png"
import ecssdviewmertric from "assets/img/docs/gettingStarted/container-insights/ecssd-view-metric.png"

<p>

>     GettingStarted / <Link to="/docs/getting-started/container-insights">Container Insights</Link> / <Link to="/docs/getting-started/container-insights#collecting-prometheus-metrics">Container Insights Prometheus</Link>

Reviewer: ```suggestion
    GettingStarted / <Link to="/docs/getting-started/container-insights">Container Insights</Link> / <Link to="/docs/getting-started/container-insights#collecting-prometheus-metrics">Container Insights foro Prometheus Support</Link>
```

</p>

## Overview

To collect Prometheus metrics from tasks running on ECS and send it to CloudWatch using AWS Distro for OpenTelemetry Collector (ADOT).
You can either run the collector as a sidecar or deploy the collector as its own ECS service for entire cluster.
To use the pre-built dashboard in CloudWatch, you need to run a single collector for the entire cluster to attach task and service metadata.
ECS tasks with Prometheus endpoints are discovered using extension [ecsobserver](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/extension/observer/ecsobserver).
(Unlike EKS, there is no builtin discovery for ECS [inside prometheus](https://github.com/prometheus/prometheus/tree/main/discovery/aws))

## Steps

> Follow these steps to run the collector and Prometheus workload on ECS. You can skip some of them if you already have clusters or tasks running.

Reviewer: ```suggestion
Follow these steps to run the ADOT Collector and Prometheus workload on ECS. You can skip some of them if you already have clusters or tasks running.
```

- Create an ECS cluster based on <Link to="/docs/setup/ecs/create-cluster">tutorial</Link> or follow Appendix to [Create ECS cluster with EC2 instances using ecs-cli](#create-ecs-cluster-with-ec2-instances)
- Create <Link to="/docs/setup/ecs/create-iam-role">IAM role</Link> and configure [IAM policy](#iam-policy). Discovery tasks requires extra ECS policies.
- Create [Configuration](#configuration) as an <Link to="/docs/setup/ecs/config-through-ssm">SSM parameter</Link>
- Create [ECS Task Definition](#ecs-task-definition)

> - Deploy [collector](#deploy-collector) as a replica service with 1 replica

Reviewer: ```suggestion
- Deploy [ADOT Collector](#deploy-collector) as a replica service with 1 replica
```

- Deploy [sample workloads](#sample-workloads)
- [View logs and metrics in the CloudWatch console](#view-logs-and-metrics-in-the-cloudwatch-console)
- [View pre-built CloudWatch dashboards](#view-pre-built-cloudwatch-dashboards)

## Permission

For discovering ECS tasks, you need to [config IAM policy](#iam-policy). For scraping discovered targets, you need to
[configure network](#network)

### IAM Policy

Extra IAM policies are required to discover ECS tasks with Prometheus metrics.
The full document with  <Link to="/docs/setup/ecs/create-iam-policy">existing polices</Link> is included in the appendix [IAM Policy Document](#iam-policy-document).
NOTE: The `EC2` policy is required if you use ECS EC2 with `bridge` network mode. You can remove it if all your tasks are fargate or uses `awsvpc` network mode.

```text
ec2:DescribeInstances
ecs:ListTasks
ecs:ListServices
ecs:DescribeContainerInstances
ecs:DescribeServices
ecs:DescribeTasks
ecs:DescribeTaskDefinition
```

### Network

> For the collector to scrape metrics, connectivity between ECS tasks using private ip is required.

Reviewer: ```suggestion
For the ADOT Collector to scrape metrics, connectivity between ECS tasks using private ip is required.
```

  awsemf:
    namespace: ECS/ContainerInsights/Prometheus # Use the exact namespace for builtin dashboard to work
    log_group_name: "/aws/ecs/containerinsights/my-ecs-cluster/prometheus" # Log group name format is fixed as well, the only variable is cluster name
    dimension_rollup_option: NoDimensionRollup
    metric_declarations:
      # NGINX
      - dimensions: [ [ ClusterName, TaskDefinitionFamily, ServiceName ] ] # dimension names are same as our relabeled keys.
        label_matchers:
          - label_names:
              - ServiceName
            regex: '^.*nginx-service$'
        metric_name_selectors:
          - "^nginx_.*$"
      # JMX
      - dimensions: [ [ ClusterName, TaskDefinitionFamily, area ] ]
        label_matchers:
          - label_names:
              - Java_EMF_Metrics # Label in metrics comes from docker label in task definition
            regex: ^true$
        metric_name_selectors:
          - "^jvm_memory_bytes_used$"

service:
  extensions: [ ecs_observer ] # extension is declared outside pipelines
  pipelines:
    metrics:
      receivers: [ prometheus ]
      processors: [ resource, metricstransform ]
      exporters: [ awsemf ]
```

> ## Deploy Collector

Reviewer: ```suggestion
## Deploy ADOT Collector
```

> Collector needs to run as single instance. We use the replica service because it handles task failure.

Reviewer: ```suggestion
ADOT Collector needs to run as single instance. We use the replica service because it handles task failure.
```

Replica is 1 otherwise we will have duplicated metrics.
If you have a large cluster, you can shard the scrape targets using `hashmod` or use different discovery configurations for multiple collectors.

```bash

> # Run collector as replica service on EC2

Reviewer: ```suggestion
# Run ADOT Collector as replica service on EC2
```

                "xray:GetSamplingStatisticSummaries",
                "cloudwatch:PutMetricData",
                "ec2:DescribeVolumes",
                "ec2:DescribeTags",
                "ssm:GetParameters"
                "ec2:DescribeInstances",
                "ecs:ListTasks",
                "ecs:ListServices",
                "ecs:DescribeContainerInstances",
                "ecs:DescribeServices",
                "ecs:DescribeTasks",
                "ecs:DescribeTaskDefinition",
                "ecs:DescribeInstances"
            ],
            "Resource": "*"
        }
    ]
}
```

### ECS Task Definition

NOTE: You need to replace placeholder values with actual value

- The account id in the example is `123456` and is used in ARN for taskRole and executionRole, replace it with your own AWS account id.
- `amazon/aws-otel-collector:v0.11.0` is using dockerhub (potential throttle) and you can use `latest` tag if you don't want to lock version.
- `/aoc/ecs-prometheus` is the name of <Link to="/docs/setup/ecs/config-through-ssm">SSM parameter</Link>.
- `networkMode` can be `bridge`, `host`, `awsvpc`. For fargate, it has to be `awsvpc`.

> - `logConfiguration` is for log from collector itself. Log groups for metrics/log collected by collector is configured using `AOT_CONFIG_CONTENT`.

Reviewer: ```suggestion
- `logConfiguration` is for log from ADOT Collector itself. Log groups for metrics/log collected by collector is configured using `AOT_CONFIG_CONTENT`.
```

> - `requiresCompatibilities` includes both EC2 and Fargate because collector is scraping metrics via HTTP and does not need special host level access.

Reviewer: ```suggestion
- `requiresCompatibilities` includes both EC2 and Fargate because ADOT Collector is scraping metrics via HTTP and does not need special host level access.
```
